### ISSUE TITLE
CPU 100% causing logins to fail, slow syncing

### ISSUE THREAD

**Issue by Redth** [realm/realm-object-server#228]
## Goals

Lower CPU load, to figure out how to diagnose this issue...

This is a production server right now and the 1 star reviews are piling up as this goes unsolved... I'd really appreciate some help.  At this point i'm almost even considering writing my own backend server to replace realm, but i'd much rather not!

## Expected Results

Responsive server

## Actual Results

CPU spikes to 100% for long periods of time, making the server unresponsive to login requests.

I'm running on Azure, I've tried increasing the VM size to something quite beefy.  I've got about 2500 realms and not a lot of data in them (most are 50-200kb, a hand full are 1-3mb), not a lot of active users at any time (maybe 100 max simultaneously).

I'm not even sure how to go about diagnosing or debugging this further.  It seems when I start the server at first everything is ok, then it starts spiking up to 100%, and I see a lot of realm opening/closings on startup. 

## Version of Realm and Tooling

- Realm Object Server Version: 1.8.1
- Flavor:
  - [X] Developer
  - [ ] Professional
  - [ ] Enterprise
- Server OS & Version: ? Ubuntu
- Client SDK Version: Xamarin 1.4.0
- Client OS & Version: iOS / Android

## Logs

I have a massive log file zipped up here: https://www.dropbox.com/s/hw2z589lxa01usm/ros.zip?dl=0

**Comment by Redth** [comment]
Here's some lldb info if it helps at all:
https://gist.github.com/Redth/041d3edc87806ab0c6a0fa2e031affb8

**Comment by Redth** [comment]
Also here's my custom auth provider: https://gist.github.com/Redth/ead6d8cf7d6bd599942ca432608e83bc

I did try disabling the custom auth provider from the configuration.yml and it didn't seem to change the 100% CPU at all...

**Comment by Redth** [comment]
Here's a strace summary running for awhile:

**Comment by morten-krogh** [comment]
@Redth I am looking into your log file now. Let us see if it reveals the problem.

**Comment by Redth** [comment]
@morten-krogh would love any updates as you're able to provide... Anxiously waiting as more 1 star app reviews roll in :( . thanks!

**Comment by morten-krogh** [comment]
@Redth I am looking. Something I see is that some internal book keeping about permissions
is taking a long time on the server after startup but that work stops after three minutes. 
Right after server startup, you might see this work reported as high CPU usage.

I am searching for more information.

I can also see that there are users whose access tokens have expired when they finally get 
through. This adds extra delay.

Can I ask why users are dissatisfied. Is it because they cannot sync their data, or because they cannot log in.

Can I also ask why the server has been restarted so often? Is that done by you on purpose?

**Comment by morten-krogh** [comment]
It seems like the server is restarted every 10 minutes.

**Comment by Redth** [comment]
I've manually restarted it some times to try different things with strace and such...

The cpu usage isn't actually hitting 100% until after the first 3 minutes. 

When it goes to 100% login does not work reliably. So thats the main issue. It seems like once they are logged in syncing is ok. 

In any case 100% CPU does not seem normal to me.

**Comment by Redth** [comment]
I'd be happy to provide a backup for you to test locally. Not sure it will show the same symptoms without users connecting though. Let me know if there is any other debugging I can provide. 

I found it odd that a lot of time seems spent in getting the datetime according to strace? That seems maybe like a timer issue on node?

@morten-krogh ^

**Comment by morten-krogh** [comment]
We have had slow login due to calculations of a Bcrypt hash function.
This takes 2-3 seconds. If multiple users log in simultaneously, this could lead to long delays.
We have made that much faster in an unreleased version.

Could this be it?

**Comment by morten-krogh** [comment]
Can you point me to a time point in your log file around which the server had 100% CPU usage?

**Comment by Redth** [comment]
It could be the bcrypt hash.

**Comment by Redth** [comment]
It does seem to happen when users are logging in.

**Comment by morten-krogh** [comment]
I see that there are errors in the JSON parsing of the access tokens. We have seen this error before. I believe that it was due to a bug in the Android client, and that it has been fixed in a new version.

Here is the log output:

2017-07-02T18:46:02.070Z - warn: auth.custom/tfp: Handle request failed with: SyntaxError: Unexpected end of JSON input
2017-07-02T18:46:02.071Z - warn: [services] internal error: SyntaxError: Unexpected end of JSON input
    at Object.parse (native)
    at request.catch.then.res (/etc/realm/custom/tfp.js:26:24)
    at process._tickCallback (internal/process/next_tick.js:109:7).

**Comment by Redth** [comment]
When. I woke up and checked early this morning CPU usage seemed normal. Then as people presumably woke up and started logging in it spiked. So I wouldn't be surprised if this is at least part of the issue

**Comment by Redth** [comment]
Is this fixed in the xamarin android client v1.5.0? If so I can push an update with that

**Comment by morten-krogh** [comment]
@Redth 

what about your own custom auth? Doesn't it circumvent our password auth?

 We will give you the newest version of ROS, so you can try it.

**Comment by morten-krogh** [comment]
So, we are talking about two issues now.

The JSON error.
The bcrypt hash function.

I will find out exactly when they were fixed.

**Comment by Redth** [comment]
Our custom auth I attached as a gist in this issue. It returns an authenticated user ID

**Comment by nirinchev** [comment]
Are users using the custom auth or the built in username/password authentication?

@morten-krogh the JSON error you're referring to was part of the native Android binding. The Xamarin binding which @Redth is using doesn't have json-related known issues, so those are either sporadic or a bug we're not aware of.

**Comment by morten-krogh** [comment]
Exactly, which seems like the bcrypt issue might not be the explanation, since our
password code does not run when you use your own authentication.

**Comment by morten-krogh** [comment]
@Redth @nirinchev Are users just using the Xamarin client here?

In that case we might have found a new error. There are lots of lines like this:

2017-07-02T18:49:01.903Z - warn: auth.custom/tfp: Handle request failed with: SyntaxError: Unexpected token u in JSON at position 0
2017-07-02T18:49:01.905Z - warn: [services] internal error: SyntaxError: Unexpected token u in JSON at position 0
    at Object.parse (native)
    at request.catch.then.res (/etc/realm/custom/tfp.js:26:24)
    at process._tickCallback (internal/process/next_tick.js:109:7).
2017-07-02T18:49:01.905Z - warn: auth.custom/tfp: Handle request failed with: SyntaxError: Unexpected token u in JSON at position 0
2017-07-02T18:49:01.906Z - warn: [services] internal error: SyntaxError: Unexpected token u in JSON at position 0
    at Object.parse (native)
    at request.catch.then.res (/etc/realm/custom/tfp.js:26:24)
    at process._tickCallback (internal/process/next_tick.js:109:7).
2017-07-02T18:49:01.906Z - warn: auth.custom/tfp: Handle request failed with: SyntaxError: Unexpected token u in JSON at position 0
2017-07-02T18:49:01.908Z - warn: [services] internal error: SyntaxError: Unexpected token u in JSON at position 0
    at Object.parse (native)
    at request.catch.then.res (/etc/realm/custom/tfp.js:26:24)
    at process._tickCallback (internal/process/next_tick.js:109:7).
2017-07-02T18:49:01.908Z - warn: auth.custom/tfp: Handle request failed with: SyntaxError: Unexpected token u in JSON at position 0
2017-07-02T18:49:01.909Z - warn: [services] internal error: SyntaxError: Unexpected token u in JSON at position 0
    at Object.parse (native)
    at request.catch.then.res (/etc/realm/custom/tfp.js:26:24)
    at process._tickCallback (internal/process/next_tick.js:109:7).


They look quite similar to the Android bug.

**Comment by nirinchev** [comment]
This looks like an issue with the custom auth provider that @Redth has posted.

**Comment by Redth** [comment]
So I think that particular issue on the auth provider is fixed. If you look closer to the more recent logs I think you will not see that same error

**Comment by nirinchev** [comment]
Just to confirm, are the majority of your users using the custom authentication provider or the built-in username/password provider from ROS?

**Comment by morten-krogh** [comment]
Okay, fine. You had a problem in the JSON parsing in your provider. 
I thought it was the client that sent malformed JSON. Fine.

Back to the CPU usage. Do you have a time point in the log where the 100% CPU usage took place?

If you use our password provider, the bcrypt hash might be the explanation.

**Comment by Redth** [comment]
We only use our custom auth, Facebook and google providers. 

I will try and find a point in time, but it's almost constant

**Comment by morten-krogh** [comment]
The JSON error is here today.

2017-07-03T01:36:58.899Z - warn: [services] internal error: SyntaxError: Unexpected token u in JSON at position 0
    at Object.parse (native)
    at request.catch.then.res (/etc/realm/custom/tfp.js:28:29)
    at process._tickCallback (internal/process/next_tick.js:109:7).

**Comment by morten-krogh** [comment]
And there are other errors in the handling of accounts. Let us look into that as well.

2017-07-03T01:36:30.493Z - warn: [services] internal error: Error: Account.provider_id must be of type: string
    at Error (native)
    at inTransaction.db (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/dao/auth/realm/account.js:92:26)
    at RealmAuthDao.inTransaction (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/dao/auth/realm/base.js:50:16)
    at AccountCollection.inTransaction (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/dao/auth/realm/baseCollection.js:99:21)
    at AccountCollection.create (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/dao/auth/realm/account.js:85:17)
    at dao.inTransaction._db (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/managers/auth/manager.js:135:36)
    at Realm._db.write (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/dao/auth/realm/base.js:44:18)
    at RealmAuthDao.inTransaction (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/dao/auth/realm/base.js:43:16)
    at AuthManager.findOrCreateAccount (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/managers/auth/manager.js:134:16)
    at resolve (/usr/lib/nodejs/realm-object-server-developer/.build/src/node/managers/auth/provider/base.js:115:31).

**Comment by Redth** [comment]
Hmm ok not sure why that is. Seems like a bad response. But that may be a symptom of the high CPU usage and a request timing out. I mentioned that I've tried disabling the custom auth in my configuration and the Ros still goes to 100% CPU.  That was one of the first things I tried

**Comment by Redth** [comment]
So my provider returns null for an account Id if the login fails. Does that cause the account provider id errors?

**Comment by morten-krogh** [comment]
@Redth I know. Sorry. I just saw this problem. I am not suggesting that this 
is the reason for the 100% CPU usage.

**Comment by Redth** [comment]
Is that ok for my provider to return null? Is there a more appropriate way to handle unsuccessful auth?

**Comment by nirinchev** [comment]
Regarding the custom provider - you should throw an error to signal that a problem has been encountered. For example in your catch clause, instead of returning `null` at the end, throw:
```js
throw new problem.HttpProblem.Unauthorized({ detail: err.toString() });
```

**Comment by Redth** [comment]
@morten-krogh here's my latest logs. Ignore the provider errors. I fixed them right after capturing this log. Still 100% CPU

https://www.dropbox.com/s/iv4kbba4msec8lj/ros%20%281%29.zip?dl=0

**Comment by Redth** [comment]
@morten-krogh it's there anything else I can do to help debug this? Anything with gdb or strace? Backups?

**Comment by morten-krogh** [comment]
I will look at the log.

Maybe we can try to give me a local copy of your server setup. Is that what you mean by backup?

Then I can try to run it myself. However, the problem might only manifest itself with real clients.
But let me try myself first.

We can also try to insert extra logging messages in your server to pin down the place where the 
problem occurs. 

Let us do this first, and then proceed from there.

**Comment by Redth** [comment]
@morten-krogh i've emailed a link to a set of backups... hopefully that will help...

**Comment by Redth** [comment]
Were you able to restore that backup? You mentioned a build with more logging added to help track the issue down too. Please let me know how I can help

**Comment by morten-krogh** [comment]
@redth I was experimenting with custom auth. I also see some high CPU usages now and then. I will get to your backup very soon. I will let you know what the next step is soon.

**Comment by morten-krogh** [comment]
@Redth 
I don't see anything suspicious in the backup. 
Next step must be to monitor your server and insert a lot of debug statements at various places
in the code. Either I should get access to your server or you will have to insert code at various 
places and send the log files.

**Comment by Redth** [comment]
@morten-krogh can we take this conversation either into the realm-public slack channel or some other messaging? let me know what's convenient to you

**Comment by morten-krogh** [comment]
We can take it on slack. Is there a channel for your project?

**Comment by Redth** [comment]
nope, i'm just on the realm-public.slack.com instance, username @redth please just pm me :)

**Comment by morten-krogh** [comment]
I don't thin I am on that channel. I have sent you an email.

**Comment by morten-krogh** [comment]
I am asking in the company to arrange for a slack channel where we can communicate.
I will get back to you.

**Comment by bmunkholm** [comment]
This has been resolved via other channel. A workaround has been made while we investigate further the rootcause.

**Comment by brayanL** [comment]
@bmunkholm this is a recursive answer, there is any workaround solution for this issue?, thank you for you answer....!

**Comment by bigfish24** [comment]
@brayanL are you currently experiencing an issue?

**Comment by brayanL** [comment]
I need update a model in Local DataBase for every user that have downloaded the app, the first time that the app is running i achieved open async Realm getInstanceAsync [,](https://realm.io/docs/java/latest/#opening-a-synchronized-realm)  I have inserted local data in Model extend RealmObject, but now I must update the same model but getting data from the server. I can achieve this, with the Developer Realm Edition, or really i need update to Profesional Realm Edition. I would like to know before starting to install Realm Object Server on my server. Thanks for you answer....

**Comment by bigfish24** [comment]
@brayanL sorry I don't understand. This issue was related to CPU use, but your issue seems related to Realm file size? Can you please open a new issue and outline it there?

**Comment by aehlke** [comment]
I’d be curious to hear about the workaround that was found. As a potential customer evaluating the service, this looks very scary - unclear whether the workaround would be sufficient for us if we also ran into this.

**Comment by bigfish24** [comment]
@aehlke we routed requests directly to the sync worker (the part of Realm Object Server that handles the actual syncing) since there was an issue around permissions. We have since made improvements to the permission system and we also have dramatically improved the sync's merge algorithm. We are prepping for a 2.0 update in September, if you want early access, email me at af@realm.io.

**Comment by AlekseyKolosovskiy** [comment]
Hi!
We have similar issue. On every client login, CPU usage 100% for 3-8 seconds.
We just released our app on AppStore. And this issue is not good for us. 
When version 2.0 will be available? Will it remove this issue?

**Comment by AlekseyKolosovskiy** [comment]
Is anybody here?
I send email 2 days ago to @bigfish24, but no answer receive. Its very critical bug, users cant normal register in our app. Now 100% CPU usage up to 30 seconds on every user creation.
This is not good behaviour for the service with price $1750 in month. We as a potential customers not understand this.

**Comment by seferdemir** [comment]
Is there any news about this issue? Is the problem was solved with version 2.0?

**Comment by bigfish24** [comment]
This should be solved with 2.0, please let us know though if you run into it again.
